infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java

```diff
@@ -131,69 +141,130 @@ public LogUnitServer(ServerContext serverContext) {
     }
 
+    // TODO(Maithem): converting exceptions to message types is really superfluous,
+    // we should be able to send the exception directly
+    void handleException(Throwable ex, ChannelHandlerContext ctx, CorfuPayloadMsg msg, IServerRouter r) {
+        if (ex.getCause() instanceof WrongEpochException) {
+            WrongEpochException wee = (WrongEpochException) ex.getCause();
+            r.sendResponse(ctx, msg, new CorfuPayloadMsg<>(CorfuMsgType.WRONG_EPOCH, wee.getCorrectEpoch()));
+        } else if (ex.getCause() instanceof OverwriteException) {
+            OverwriteException owe = (OverwriteException) ex.getCause();
+            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_OVERWRITE
+                    .payloadMsg(owe.getOverWriteCause().getId()));
+        } else if (ex.getCause() instanceof DataOutrankedException) {
+            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_DATA_OUTRANKED.msg());
+        } else if (ex.getCause() instanceof ValueAdoptedException) {
+            ValueAdoptedException vae = (ValueAdoptedException) ex.getCause();
+            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_VALUE_ADOPTED.payloadMsg(vae.getReadResponse()));
+        } else if (ex.getCause() instanceof TrimmedException) {
+            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_TRIMMED.msg());
+        } else {
+            throw new LogUnitException(ex);
+        }
+    }
+
     /**
      * Service an incoming write request.
      */
     @ServerHandler(type = CorfuMsgType.WRITE)
     public void write(CorfuPayloadMsg<WriteRequest> msg, ChannelHandlerContext ctx, IServerRouter r) {
         log.debug("log write: global: {}, streams: {}", msg.getPayload().getToken(),
                 msg.getPayload().getData().getBackpointerMap());
-
-        try {
-            LogData logData = (LogData) msg.getPayload().getData();
-            logData.setEpoch(msg.getEpoch());
-            dataCache.put(msg.getPayload().getGlobalAddress(), logData);
-            r.sendResponse(ctx, msg, CorfuMsgType.WRITE_OK.msg());
-
-        } catch (OverwriteException ex) {
-            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_OVERWRITE.payloadMsg(ex.getOverWriteCause().getId()));
-        } catch (DataOutrankedException e) {
-            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_DATA_OUTRANKED.msg());
-        } catch (ValueAdoptedException e) {
-            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_VALUE_ADOPTED.payloadMsg(e
-                    .getReadResponse()));
-        }
+        LogData logData = (LogData) msg.getPayload().getData();
+
+        batchWriter.addTask(WRITE, msg)
+                .thenAccept(Void -> {
+                    dataCache.put(msg.getPayload().getGlobalAddress(), logData);
+                    r.sendResponse(ctx, msg, CorfuMsgType.WRITE_OK.msg());
+                }).exceptionally(ex -> {
+                    handleException(ex, ctx, msg, r);
+                    return null;
```

Comment:
I don't like that we have to return null. Can you please replace `exceptionally()` by `handle(...)` or `whenCompleted(...)`

Reply:
I agree, I don't like it too, but I believe that's a shortcoming of the `CompletableFuture` API, not the usage. I see the application of `exceptionally` appropriate here because its used as the last stage. Furthermore, I don't see how `handle` and `whenCompleted` are a cleaner alternatives, because using them will also require "null checking".